### PR TITLE
Fix LastBatchPolicy doc and update Parallel ES wording

### PR DESCRIPTION
**Pull request by klecki**
Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

## Description
- [ ] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [x] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR
Update the Parallel External source error regarding CUDA context 
based on user feedback - mention DALI and when the context can be
acquired.

Expose LastBatchPolicy as enum in the docs, so the link works,
fix the layout of docs around a bit.

#### Additional information
- Affected modules and functionalities:

- Key points relevant for the review:

### Tests
- [ ] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [x] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [x] Docstring
  - [ ] Doxygen
  - [x] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2198

### CODE REVIEW

**Comment by klecki** [comment]
!build

File: docs/advanced_topics_sharding.rst [hunk 1]
```diff
@@ -17,50 +17,61 @@ The parameter also ensures that all pipelines return the same number of batches,
...
-  the shard size is divisible by the batch size.
 - | ``reader_name`` - Allows you to provide the name of the reader that drives the iterator and
    provides the necessary parameters.
 
   .. note::
-    We recommend that you use this option. With this option, the next two options are excluded and
-    cannot be used.
+    We recommend that you use this option. With this option, the next two options
+    (``size`` and ``last_batch_padded``) are obtained autmatically from pipeline configuration
+    and should not be provided manually for the iterator.
```
Comment (jantonguirao) [inline docs/advanced_topics_sharding.rst:34]:
```suggestion
    and should not need to be provided explicitly to the iterator.
```

File: docs/advanced_topics_sharding.rst [hunk 2]
```diff
@@ -17,50 +17,61 @@ The parameter also ensures that all pipelines return the same number of batches,
...
+  floor(((id + epoch_num) % num_shards + 1) * dataset_size / num_shards) - floor(((id + epoch_num) % num_shards) * dataset_size / num_shards)
 
 When the second formula is used, providing a size value once at the beginning of the training works
 only when the ``stick_to_shard`` reader option is enabled and prevents DALI from rotating shards.
 When this occurs, use the first formula.
 
-To address these challenges, use the ``reader_name`` parameter and allow the iterator
-handle the details.
+To address these challenges, use the ``reader_name`` parameter and allow the iterator to
+handle the confguration autmatically.
```
Comment (jantonguirao) [inline docs/advanced_topics_sharding.rst:76]:
```suggestion
handle the configuration automatically.
```

File: docs/advanced_topics_sharding.rst [hunk 3]
```diff
@@ -17,50 +17,61 @@ The parameter also ensures that all pipelines return the same number of batches,
...
   | This option is more flexible and accurate and takes into account that shard size for a pipeline
     can differ between epochs when the shards are rotated.
-- ``size``: Displays the size of the shard for an iterator or, if there is more than one shard,
+- ``size``: Provides the size of the shard for an iterator or, if there is more than one shard,
   the sum of all shard sizes for all wrapped pipelines.
 - | ``last_batch_padded``: Determines whether the tail of the data consists of data from the next
     shard (``False``) or is duplicated dummy data (``True``).
   | It is applicable when the shard size is not a multiple of the batch size,
-- | ``last_batch_policy`` - Determines the police about the last batch when the shard size is not
+- | ``last_batch_policy`` - Determines the policy about the last batch when the shard size is not
```
Comment (mzient) [inline docs/advanced_topics_sharding.rst:43]:
```suggestion
- | ``last_batch_policy`` - Determines the handling of the last batch when the shard size is not
```

File: docs/advanced_topics_sharding.rst [hunk 4]
```diff
@@ -17,50 +17,61 @@ The parameter also ensures that all pipelines return the same number of batches,
...
     divisible by the batch size.
-  | Only partially filled with the data or dropped entirely if it. The possible options are:
+  | It affects batches only partially filled with the data. It is controlled by the
+    :meth:`~nvidia.dali.plugin.base_iterator.LastBatchPolicy` enum.
```
Comment (mzient) [inline docs/advanced_topics_sharding.rst:46]:
```suggestion
  | It affects only the batches partially filled with data. See :meth:`~nvidia.dali.plugin.base_iterator.LastBatchPolicy` enum for possible values.
```
An option is not controlled by its type.

File: docs/advanced_topics_sharding.rst [hunk 5]
```diff
@@ -17,50 +17,61 @@ The parameter also ensures that all pipelines return the same number of batches,
...
+
+- ``fill_last_batch`` – (Deprecated in favour of ``last_batch_policy``) Determines whether the last
+  batch should be full, regardless of  whether the shard size is divisible by the batch size.
```
Comment (mzient) [inline docs/advanced_topics_sharding.rst:49]:
```suggestion
- ``fill_last_batch`` – (Deprecated in favour of ``last_batch_policy``) Determines whether the last
  batch should be full, regardless of whether the shard size is divisible by the batch size.
```

Comment (mzient) [inline docs/advanced_topics_sharding.rst:49]:
```suggestion
- ``fill_last_batch`` – (Deprecated in favour of ``last_batch_policy``) Determines whether the last
  batch should be full, regardless of whether the shard size is divisible by the batch size.
```
remove extra space

File: dali/python/nvidia/dali/_multiproc/pool.py [hunk 6]
```diff
@@ -177,11 +177,15 @@ def __init__(
             raise RuntimeError("Chunk capacity must be positive integer")
         if start_method == 'fork' and _b.HasCudaContext():
             raise RuntimeError(
-                "Cannot fork process when there is CUDA context bound to it. "
-                "Make sure you build pipeline before CUDA context is acquired or change Python workers "
-                "starting method from ``fork`` to ``spawn`` (see Piepline's ``py_start_method`` option). "
+                "Error when starting Python worker threads for DALI parallel External Source. "
+                "Cannot fork a process when there is CUDA context already bound to the process. "
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/python/nvidia/dali/_multiproc/pool.py:181]: done

File: dali/python/nvidia/dali/_multiproc/pool.py [hunk 7]
```diff
@@ -177,11 +177,15 @@ def __init__(
...
+                "CUDA context is acquired during ``Pipeline.build()`` or can be acquired by other "
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/python/nvidia/dali/_multiproc/pool.py:182]: done

File: dali/python/nvidia/dali/_multiproc/pool.py [hunk 8]
```diff
@@ -177,11 +177,15 @@ def __init__(
...
+                "library that interacts with CUDA, for example DL framework creating CUDA tensors."
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/python/nvidia/dali/_multiproc/pool.py:183]: done, I prefer `that interacts`

File: dali/python/nvidia/dali/plugin/base_iterator.py [hunk 9]
```diff
@@ -30,13 +30,12 @@ def _iterator_deprecation_warning():
...
-
-        FILL - Fills the last batch with the data wrapping up the data set. The precise
-               behavior depends on the reader which may duplicate the last sample to fill the batch
-        DROP - If the last batch cannot be fully filled by the data from given epoch it is dropped
-        PARTIAL - Returns the part of the last batch filled with the data relevant to given epoch
+    Describes the last batch policy behavior when there are not enough samples in the epoch
+    to fill a whole batch.
+        * FILL - Fills the last batch with the data wrapping up the data set. The precise
+          behavior depends on the reader which may duplicate the last sample to fill the batch
+        * DROP - If the last batch cannot be fully filled by the data from given epoch it is dropped
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/python/nvidia/dali/plugin/base_iterator.py:37]: done

File: dali/python/nvidia/dali/plugin/base_iterator.py [hunk 10]
```diff
@@ -30,13 +30,12 @@ def _iterator_deprecation_warning():
...
+        * PARTIAL - Returns the part of the last batch filled with the data relevant to given epoch
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/python/nvidia/dali/plugin/base_iterator.py:38]: done

File: docs/advanced_topics_sharding.rst (hunk 1 above)
Comment (klecki) [reply docs/advanced_topics_sharding.rst:34]: done, split the sentece.

File: docs/advanced_topics_sharding.rst (hunk 3 above)
Comment (klecki) [reply docs/advanced_topics_sharding.rst:43]: done

File: docs/advanced_topics_sharding.rst (hunk 4 above)
Comment (klecki) [reply docs/advanced_topics_sharding.rst:46]: done

File: docs/advanced_topics_sharding.rst (hunk 5 above)
Comment (klecki) [reply docs/advanced_topics_sharding.rst:49]: done

File: docs/advanced_topics_sharding.rst (hunk 2 above)
Comment (klecki) [reply docs/advanced_topics_sharding.rst:76]: done

File: dali/python/nvidia/dali/plugin/base_iterator.py [hunk 11]
```diff
@@ -30,13 +30,12 @@ def _iterator_deprecation_warning():
...
-        Describes the last batch policy behavior when there are not enough samples in the epoch
-        to fill a whole batch.
-
-        FILL - Fills the last batch with the data wrapping up the data set. The precise
-               behavior depends on the reader which may duplicate the last sample to fill the batch
-        DROP - If the last batch cannot be fully filled by the data from given epoch it is dropped
-        PARTIAL - Returns the part of the last batch filled with the data relevant to given epoch
+    Describes the last batch policy behavior when there are not enough samples in the epoch
+    to fill a whole batch.
+        * FILL - Fills the last batch with the data wrapping up the data set. The precise
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/python/nvidia/dali/plugin/base_iterator.py:35]: done

File: docs/advanced_topics_sharding.rst [hunk 12]
```diff
@@ -17,50 +17,61 @@ The parameter also ensures that all pipelines return the same number of batches,
...
 
-- ``fill_last_batch`` – Determines whether the last batch should be full, regardless of whether
-  the shard size is divisible by the batch size.
 - | ``reader_name`` - Allows you to provide the name of the reader that drives the iterator and
    provides the necessary parameters.
 
   .. note::
-    We recommend that you use this option. With this option, the next two options are excluded and
-    cannot be used.
+    We recommend that you use this option. With this option, the next two options
```
[the remark this answers is not in the document]
Comment (klecki) [reply docs/advanced_topics_sharding.rst]: done

File: docs/advanced_topics_sharding.rst [hunk 13]
```diff
@@ -17,50 +17,61 @@ The parameter also ensures that all pipelines return the same number of batches,
...
+    (``size`` and ``last_batch_padded``) are obtained autmatically from pipeline configuration
```
[the remark this answers is not in the document]
Comment (klecki) [reply docs/advanced_topics_sharding.rst]: done

**Comment by klecki** [comment]
!build

**Review by jantonguirao — APPROVED** [review]

**Review by awolant — APPROVED** [review]

**Pull request merged**